syntax = "proto3";
package Protocol;

import "Enum.proto";
import "Struct.proto";

message C_LOGIN
{
	string name = 1;
}

message C_CONSUME_ITEM
{
	uint64 item_id = 1;
	uint64	tab_type = 2;
	uint64	slot_index = 3;
}

message C_DROP_ITEM
{
	uint64	item_id = 1;
	uint64	tab_type = 2;
	uint64	slot_index = 3;
	uint64	quantity = 4;
}

message C_MOVE_INVENTORY_ITEM
{
	uint64 from_tab = 1;
	uint64 from_index = 2;
	uint64 to_tab = 3;
	uint64 to_index = 4;
}

message C_EQUIP
{
	uint64	item_id = 1;
	uint64	tab_type = 2;
	uint64	slot_index = 3;
}

message C_UNEQUIP
{
	uint64 item_id = 1;
	uint64 slot_type = 2;
}

message C_SORT_INVENTORY 
{
	uint64 tab_type = 1;
}

message C_MOVE
{
	uint64 direction = 1;
	int64 move_time = 2;
}
message C_CHAT
{
	string msg = 1;
}
message C_TELEPORT
{
	uint64 id = 1;
}
message C_ATTACK
{
	uint64 id = 1;
	uint64 skill = 2;
}
message C_LOGOUT
{
	uint64 id = 1;
}

// server -> client
message S_LOGIN
{
	bool success = 1;
	repeated User players = 2;
	repeated Stat stats = 3;
	S_LOAD_INVENTORY inventory = 4;
	S_LOAD_EQUIPMENT equipment = 5;
}

message S_LOAD_INVENTORY
{
	repeated InventorySlot inventory = 1;
}

message S_LOAD_EQUIPMENT
{
	repeated EquipmentItem equipment = 1;
}

message S_CONSUME_RESULT {
	bool success = 1; 
	uint64 item_id = 2;
	uint64 tab_type = 3;
	uint64 slot_index = 4;
	uint64 new_quantity = 5; 
}
message S_DROP_RESULT {
	bool success = 1;
	uint64 item_id = 2;
	uint64 tab_type = 3;
	uint64 slot_index = 4;
	//uint64 quantity_dropped = 5;
	//uint64 remaining_quantity = 6;
}

message S_MOVE_INVENTORY_RESULT {
	bool success = 1;
	uint64 from_tab = 2;
	uint64 from_index = 3;
	uint64 to_tab = 4;
	uint64 to_index = 5;
	uint64 moved_item_id = 6;
	uint64 quantity = 7;
}

message S_EQUIP_RESULT {
	bool success = 1;
	uint64 item_id = 2;
	uint64 slot_type = 3; 
}

message S_UNEQUIP_RESULT {
	bool success = 1;
	uint64 item_id = 2;
	uint64 slot_type = 3;
	uint64 to_tab_type = 4;   
	uint64 to_slot_index = 5;   // 인벤으로 돌아갈 슬롯
}

message S_GOLD_CHANGE {
	uint64 new_gold = 1;    // 변경 후 골드
	int64 delta = 2;        // 증감량 (양수/음수 모두 가능)
}

message S_ADD_OBJECT
{
	repeated AddPlayer players = 1;
}

message S_REMOVE_OBJECT
{
	uint64 id = 1;
}

message S_MOVE_OBJECT
{
	repeated Move moves = 1;
}

message S_CHAT
{
	repeated Mess messages = 1;
}

message S_STAT_CHANGE
{
	repeated Stat stats = 1;
}

message S_DAMAGE
{
	uint64 damage = 1;
}

message S_RESPAWN
{
	uint64 hp = 1;
	uint64 x = 2;
	uint64 y = 3;
	uint64 exp = 4;
}